### DISCUSSION TITLE
Images from public folder not loading in dynamic path

### DISCUSSION THREAD

**Discussion by OgLuka** [vercel/next.js#38382] [Help] [answered]
Hi guys , I already created and described my problem on StackOverflow and no one is answering yet , pls check  https://stackoverflow.com/questions/72886839/next-js-image-from-public-folder-not-loading-on-dynamic-paths

**Comment by icyJoseph** [comment]
Hi, 

A few things:

- You can use images located in the public folder by just doing:
```js
<img src="/image-name.png" alt="..." />
```

Next.js knows that static assets are in the public file, so you can skip saying `/public/image-name`, please refer to this page, https://nextjs.org/docs/basic-features/static-file-serving.

- If you want to have the SVG files as React components you can do that, by using `svgr` as documented here, https://react-svgr.com/docs/next/


Now into the issue, do you have a middleware file? What Nextjs version are you using? I tried with 12.2.0, and it works fine. Do you have a reproduction repo?

**Reply by OgLuka** [reply]
Hi, thanks for answering me. 
first of all, I tried giving it src directly and not as react component but then the <Image> tag asked me to have width and height or layout="fill" so it broke my design. 
Second, when I am importing images as Components without /public/ in front it says it couldn't; find it.
Sorry, but what is the middleware file? and I'm using the next 12.1.6 . lastly no I don't have a public repo for this project yet. I could create it tho. Thanks again.

**Reply by icyJoseph** [reply]
> first of all, I tried giving it src directly and not as react component but then the  tag asked me to have width and height or layout="fill" so it broke my design.

Ok, but this is something you could fix with CSS and a wrapper `div`, element.

> Second, when I am importing images as Components without /public/ in front it says it couldn't; find it.

To import them using `import x from 'path';`, then you need to add the public part, but that shouldn't be necessary.

Ok, so no middleware, because a middleware file could rewrite the request. I tested with 12.2.0, I'll try it out with 12.1.6.

**Reply by OgLuka** [reply]
Yes i also have wrapper but anyways i will try if giving src path without importing react component will work and get back to u , but the strange thing for me is that the problem happens only in the dynamic route and i wanted to know why is that.

**Reply by icyJoseph** [reply]
Mmm nope, can't get it to break.

Please try to reproduce as small as possible. You can use the Next.js template from https://stackblitz.com/, and share that, it'll make it faster.

**Reply by icyJoseph** [reply]
> only in the dynamic route and i wanted to know why is that.

I also want to know what's going on there, but I can't get it to break.

**Reply by OgLuka** [reply]
Okay i will try recreating it here , Thanks for your time. i'll post it here when i'm done.

**Reply by icyJoseph** [reply]
No worries, we'll figure this out!

**Reply by zourdyzou** [reply]
@icyJoseph I kinda have the same issue, and you mentioned in the above statement that the middleware might also cause it? Could you please explain to me what is the best way to fix it?

**Comment by OgLuka** [comment]
Sorry for the wait. so i recreated it here and it worked somehow idk why https://stackblitz.com/edit/nextjs-cavc5g?file=pages%2F_app.js,components%2FNavbar.js,pages%2Fsomething%2Fsomething1%2F[id].js,pages%2Findex.js,package.json
so because i still don't know what the problem is i created public repo and uploaded my project . it would mean the world if u could check it.
https://github.com/OgLuka/NextTest.git

**Comment by icyJoseph** [comment, marked as answer]
Hi

The issue is the `path` used for images in the Next.js config file.

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  images: {
    loader: "akamai",
    path: "", // <----- THIS IS THE ISSUE
  },
};

module.exports = nextConfig;
```

You should have the path set to `/`

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  images: {
    loader: "akamai",
    path: "/",
  },
};

module.exports = nextConfig;
```

Why does it happen only on dynamic routes? Well it doesn't just happen in dynamic routes. It happens because of the path depth.

If you create `pages/register/foo.js`, the same issue happens.

The above is key because, when the `path` in `next.config.js` is set to `''`, the paths become relative to the parent of the current page, AKA, as the current directory. If you are in `http://localhost:3000/register`, then the current directory is `/`, once we dive one step in depth, into register `http://localhost:3000/register/foo`, now the current directory is `register`, and there's no SVG's there.

> https://stackoverflow.com/a/7613355

Using the `path: '/'`, ensures that URLs are absolute, that is, relative to the root of the site.

## Details of implementation

If you must know why:

The akamai loader looks like:

```js
function normalizeSrc(src: string): string {
  return src[0] === '/' ? src.slice(1) : src
}

function akamaiLoader({
  config,
  src,
  width,
}: ImageLoaderPropsWithConfig): string {
  return `${config.path}${normalizeSrc(src)}?imwidth=${width}`
}
```

When the `src` starts with `/`, the normalise function removes it, and then the akamai loader, joins the `path` (empty string), with the src without leading slash and the query. Now the path is relative to the current directory, and not to the root of the project. The fix is to include a path equal to `/`.


### Bonus

There's this piece of JSX in the NavBar:

```js
  <Link href="/abiplatform">
         <Image src={SquareLogo} alt="brandergate square logo" />
  </Link>
```

Wrap the image with an anchor tag.

**Reply by OgLuka** [reply]
Understood everything thanks so much, so why did my next create image without "/" in it? or it is not default. anyways thanks again for your time , i'll change this and continue working on my project.

**Reply by icyJoseph** [reply]
It's because the config was telling it to not use the default loader, but the akamai, and the path was set to `empty` string.

**Reply by OgLuka** [reply]
Oh okay!

**Comment by bucky0112** [comment]
```
const nextConfig = {
  reactStrictMode: true,
  images: {
    loader: "akamai",
    path: "/",
  },
};
```
When I set `/` to path, it will show public image in dynamic path, but image url from fetch API, also plus `/` on image src.

![CleanShot 2022-10-12 at 16 07 14](https://user-images.githubusercontent.com/49245415/195286758-6d4a1129-5b5e-4382-8934-96b52b58d3ff.png)

**Reply by hamonCordova** [reply]
Some solution? Same here

**Reply by hamonCordova** [reply]
The solution that i had is removing '/' from path in images nextConfig. And add a '/' + image.src on all public assets. This way, all the external and internal images will be rendered correctly. 
It's boring to do it, but was my solution.

**Comment by bucky0112** [comment]
I’m forgot how to fix it, but I give up use path.

￼

**Comment by ifismailkhan** [comment]
Next.js can serve static files, like images, under a folder called public in the root directory. Files inside public can then be referenced by your code starting from the base URL (/).

For example, if you add me.png inside public, the following code will access the image:

`import Image from 'next/image'
 
export function Avatar() {
  return <Image src="/me.png " alt="me" width="64" height="64" />
}`

**Comment by Pnlvfx** [comment]
Next app router has a serious problem in production when you try to access the image from the public folder from the url, it just fall to 404. I tried to report it but they close the issue. Try it!

**Reply by daviancode** [reply]
Yes, that's absolutely right.
When I deploy to the production client, the public path cannot reload new files.
Please, anyone, help me!

**Reply by jean1190** [reply]
It's as intended unfortunately. It's static and not dynamic.

**Reply by Wimmind** [reply]
I have the same problem, but even in development mode, if I specify src='file.png' in the Image, and it will lie on the path public/file.png, then the error 404 not found occurs. help!!!

**Reply by Pnlvfx** [reply]
> I have the same problem, but even in development mode, if I specify src='file.png' in the Image, and it will lie on the path public/file.png, then the error 404 not found occurs. help!!!

This is not the same problem.
You have to write src='/file.png'.

**Reply by daviancode** [reply]
finally, i'm using api to render image in Production

src='/api?filename=file.jpeg'

```
import fs from 'fs';
import { NextRequest, NextResponse } from "next/server";

export async function GET(req: NextRequest) {
    try {
        const filename: any = req.nextUrl.searchParams.get("filename");
        const blob = fs.readFileSync(`${process.cwd()}/public/${filename}`);
        return new NextResponse(blob, {
            headers: {
                'Content-Type': 'image/jpeg',
            }
        })
    } catch (error) {
        return NextResponse.json({ error: error })
    }
}
```

**Comment by Pnlvfx** [comment]
No that’s not what I mean. Using next is canary latest release and Vercel I wasn’t able to access images from the browser, in development my images were correctly shown, but in production a 404 appears when I tried to access the images from browser. I didn’t make much debugging because I was in production and so I had to switch to cloudinary.
However on my side I didn’t make a lot of stuff that could have broken the public dir, so if you can take a look at it and tell me if was my issue or if there is an issue related.
Or maybe vercel?
Because in development everything works fine!

If you don’t understand what I mean, I can access the images using the next image component, but then it happens that I use an image as og_image and when I went to check if it was correctly displaying on the browser, it wasn’t instead there was a 404 error with some strange messages, I also had open an issue but was closed as it was without repro.

Sorry I’m not English, I hope you understand my issue.

Inviato da iPhone

On Aug 22, 2024, at 6:24 PM, jean1190 ***@***.***> wrote:

﻿

It's as intended unfortunately. It's static and not dynamic.

**Reply by jean1190** [reply]
I thought I had the exact same problem. The thing is, the "public" path cannot reload new files in prod even if it works in dev.

Sorry if I missunderstood your issue.

**Comment by Pnlvfx** [comment]
I didn’t change it in production
Inviato da iPhone

On Aug 22, 2024, at 6:43 PM, jean1190 ***@***.***> wrote:

﻿

I had the exact same problem. The thing is, the "public" path cannot reload new files in prod even if it works in dev.

**Comment by fadilkun45** [comment]
any update for this problem ?  in nextJS 15 its same too i tried config next.config.ts but nothing happened 

here of my last config

import type { NextConfig } from "next";

```
const nextConfig: NextConfig = {
  basePath: "/tm"
  output: "standalone",
  images: { unoptimized: true, path:  "/" },
  eslint: {
    ignoreDuringBuilds: true,
  },
};

export default nextConfig;
```
